File: 백준/Gold/13023. ABCDE/ABCDE.py
import sys
input = sys.stdin.readline

n,m = map(int,input().split())
graph = [[] for _ in range(n)]
for _ in range(m):
    a,b = map(int,input().split())
    graph[a].append(b)
    graph[b].append(a)



def dfs(v,cnt):
    if cnt == 5 :
        print(1)
        exit()

    for i in graph[v] :
        if visited[i] == 0 :
            visited[i] = 1
            dfs(i,cnt+1)
            visited[i] = 0

visited = [0]*n
for i in range(n):
    visited[i] = 1
    dfs(i,1)
    visited[i] = 0
print(0)

File: 백준/Gold V/13023. ABCDE/ABCDE.java
import java.util.*;
import java.io.*;

public class Main {

    static List<Integer>[] graph;
    static boolean[] visited;
    static int ans = 0;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        graph = new ArrayList[n];
        visited = new boolean[n];

        for (int i = 0; i < n; i++) {
            graph[i] = new ArrayList<Integer>();
        }

        int a; int b;
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            a = Integer.parseInt(st.nextToken());
            b = Integer.parseInt(st.nextToken());
            graph[a].add(b);
            graph[b].add(a);
        }
        for (int i = 0; i < n; i++) {
            visited[i] = true;
            dfs(n, i, 1);
            visited[i] = false;
        }

        System.out.println(ans);
    }

    private static void dfs(int n, int i, int cnt) {

        if (cnt >= 5 || ans == 1) {
            ans = 1;
            return;
        }

        for (int v : graph[i]) {
            if (!visited[v]) {
                visited[v] = true;
                dfs(n,v, cnt+1);
                visited[v] = false;
            }
        }
    }
}
